pkg/controller/ttlafterfinished/ttlafterfinished_controller.go

```diff
@@ -0,0 +1,446 @@
+	defer tc.jobsDelayCleanupQueue.Close()
+
+	glog.Infof("Starting TTL after finished controller")
+	defer glog.Infof("Shutting down TTL after finished controller")
+
+	if !controller.WaitForCacheSync("TTL after finished", stopCh, tc.jListerSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(tc.runJobCleanupWorker, time.Second, stopCh)
+	}
+	go wait.JitterUntil(tc.runJobDelayCleanupWorker, 10*time.Second, 0.2, true, stopCh)
+
+	<-stopCh
+}
+
+func (tc *Controller) addJob(obj interface{}) {
+	job := obj.(*batch.Job)
+	glog.V(4).Infof("Adding job %s/%s", job.Namespace, job.Name)
+
+	// `addJobToCleanupFn` enqueues the Job when TTL expires
+	addJobToCleanupFn := func(j *batch.Job) error {
+		tc.addJobToCleanup(j)
+		return nil
+	}
+	// Check the Job's TTL and run `addJobToCleanupFn` when the TTL expires
+	err := tc.checkTTLAndReact(job, addJobToCleanupFn)
+	if err != nil {
+		glog.V(2).Infof("Error cleaning up Job %s/%s: %v", job.Namespace, job.Name, err)
+	}
+}
+
+func (tc *Controller) updateJob(old, cur interface{}) {
+	job := cur.(*batch.Job)
+	glog.V(4).Infof("Updating job %s/%s", job.Namespace, job.Name)
+
+	// `addJobToCleanupFn` enqueues the Job when TTL expires
+	addJobToCleanupFn := func(j *batch.Job) error {
+		tc.addJobToCleanup(j)
+		return nil
+	}
+	// Check the Job's TTL and run `addJobToCleanupFn` when the TTL expires
+	err := tc.checkTTLAndReact(job, addJobToCleanupFn)
+	if err != nil {
+		glog.V(2).Infof("Error cleaning up Job %s/%s: %v", job.Namespace, job.Name, err)
+	}
+}
+
+func (tc *Controller) deleteJob(obj interface{}) {
+	job := obj.(*batch.Job)
+	glog.V(4).Infof("Deleting job %s/%s", job.Namespace, job.Name)
+
+	// Remove the deleted Job from the delay cleanup queue
+	tc.jobsDelayCleanupQueue.Delete(job)
+}
+
+func (tc *Controller) addJobToCleanup(job *batch.Job) {
+	glog.V(4).Infof("Add job %s/%s to cleanup", job.Namespace, job.Name)
+	key, err := controller.KeyFunc(job)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", job, err))
+		return
+	}
+
+	tc.jobsToCleanup.Add(key)
+}
+
+func (tc *Controller) runJobCleanupWorker() {
+	for tc.jobCleanupWorker() {
+	}
+}
+
+func (tc *Controller) jobCleanupWorker() bool {
+	key, quit := tc.jobsToCleanup.Get()
+	if quit {
+		return false
+	}
+	defer tc.jobsToCleanup.Done(key)
+
+	err := tc.cleanupJob(key.(string))
+	tc.handleErr(err, key)
+
+	return true
+}
+
+func (tc *Controller) runJobDelayCleanupWorker() {
+	for tc.handleJobsDelayCleanup() {
+	}
+}
+
+func (tc *Controller) handleJobsDelayCleanup() bool {
+	if tc.jobsDelayCleanupQueue.IsClosed() {
+		glog.V(2).Infof("The delay cleanup queue for Jobs is closed.")
+		return false
+	}
+
+	nowUTC := tc.clock.Now().UTC()
+
+	tc.nextCheckAtMutex.RLock()
+	checkAt := tc.nextCheckAt
+	tc.nextCheckAtMutex.RUnlock()
+
+	// No need to check delay cleanup queue for now.
+	if checkAt == nil || checkAt.IsZero() || nowUTC.Before(checkAt.UTC()) {
+		return true
```

Comment:
Is this just hot looping in that case?

Reply:
It's checking the time to clean up every 10s, and only continue when it's time to clean up.